src/coreclr/jit/importer.cpp

```diff
@@ -8231,6 +8231,35 @@ GenTree* Compiler::impImportStaticFieldAccess(CORINFO_RESOLVED_TOKEN* pResolvedT
         }
         break;
 
+        case CORINFO_FIELD_STATIC_DATASEGMENT:
+        {
+#ifdef FEATURE_READYTORUN
+            assert(opts.IsReadyToRun());
+            assert((pFieldInfo->fieldFlags & CORINFO_FLG_FIELD_INITCLASS) == 0);
+            
+            if (pFieldInfo->fieldLookup.accessType == InfoAccessType::IAT_VALUE)
+            {
+                op1 = gtNewIconHandleNode((size_t)pFieldInfo->fieldLookup.addr, GTF_ICON_STATIC_HDL);
+                op1->gtType = TYP_BYREF;
+            }
+            else
+            {
+                assert(pFieldInfo->fieldLookup.accessType == InfoAccessType::IAT_PVALUE);
+                op1 = gtNewIndOfIconHandleNode(TYP_BYREF, (size_t)pFieldInfo->fieldLookup.addr, GTF_ICON_GLOBAL_PTR, false);
+            }
+
+            FieldSeqNode* fs = GetFieldSeqStore()->CreateSingleton(pResolvedToken->hField);
+            op1              = gtNewOperNode(GT_ADD, TYP_BYREF, op1,
+                                new (this, GT_CNS_INT) GenTreeIntCon(TYP_INT, pFieldInfo->offset, fs));
+
+            // Prevent CSE from adding the offset to the base
+            op1->gtFlags |= GTF_DONT_CSE;
```

Comment:
Why prevent CSE unconditionally? At least for the `IAT_PVALUE` case it seems it could be valuable, for the `IAT_VALUE` less so I suppose, except on ARM/64.

Is the issue here that we cannot fold `Icon + Offset` and end up CSEing it?
(If so would be worth a comment to that effect; seems like a bug in `gtSetEvalOrder`'s logic for setting `ADDRMODE_NO_CSE` -- perhaps that's what the existing comment meant to say?)

Reply:
I was seeing RyuJIT fold the constant handle with the constant and that created an invalid handle (the handle is not an address when we're AOT compiling. It's an actual handle that should not be incremented/decremented. Could be that this is a bug in CSE.

Reply:
I see. I think I know what the issue is (it should be a VN + assertion propagation bug). Could you open a bug on this and add a `TODO-Bug: workaround for <link>` here?